crates/client/sync/src/l2.rs

```diff
@@ -44,22 +46,42 @@ pub struct L2StateUpdate {
     pub block_hash: Felt,
 }
 
-#[allow(clippy::too_many_arguments)]
-#[tracing::instrument(skip(backend, updates_receiver, block_import, validation), fields(module = "Sync"))]
-async fn l2_verify_and_apply_task(
-    backend: Arc<MadaraBackend>,
-    mut updates_receiver: mpsc::Receiver<PreValidatedBlock>,
+pub struct L2VerifyApplyConfig {
     block_import: Arc<BlockImporter>,
-    validation: BlockValidationContext,
     backup_every_n_blocks: Option<u64>,
-    telemetry: TelemetryHandle,
+    flush_every_n_blocks: u16,
     stop_on_sync: bool,
-    cancellation_token: tokio_util::sync::CancellationToken,
+    telemetry: TelemetryHandle,
+    validation: BlockValidationContext,
+    block_conv_receiver: mpsc::Receiver<PreValidatedBlock>,
+}
+
+#[tracing::instrument(skip(backend, ctx, config), fields(module = "Sync"))]
+async fn l2_verify_and_apply_task(
+    backend: Arc<MadaraBackend>,
+    ctx: ServiceContext,
+    config: L2VerifyApplyConfig,
 ) -> anyhow::Result<()> {
-    while let Some(block) = channel_wait_or_graceful_shutdown(pin!(updates_receiver.recv()), &cancellation_token).await
-    {
+    let mut last_block_n = 0;
+
+    let L2VerifyApplyConfig {
+        block_import,
+        backup_every_n_blocks,
+        flush_every_n_blocks,
+        stop_on_sync,
+        telemetry,
+        validation,
+        mut block_conv_receiver,
+    } = config;
+
+    while let Some(block) = channel_wait_or_graceful_shutdown(pin!(block_conv_receiver.recv()), &ctx).await {
         let BlockImportResult { header, block_hash } = block_import.verify_apply(block, validation.clone()).await?;
 
+        if header.block_number - last_block_n >= flush_every_n_blocks as u64 {
```

Comment:
re: i really think this should not be here, flushing should def not be the concern of l2.rs

Reply:
Where do you suggest we move this instead? This was already called indirectly in the `verify_apply` method if I remember correctly. The issue is that if we are running a full node with only sync services enabled we still need to have  a point where new blocks are flushed. Also I don't quite see how we can share information across services in an elegant way. Imo it makes most sense to call `flush` were information about the block store is directly available, be that `sync`, `block_production` or the `mempool`.

Reply:
i think it makes more sense to call it in block_import? this seems way too low level to be a concern of l2.rs

Reply:
what do you mean sharing info in this case?

Reply:
Tangentially related but I have added a timer inside the l2 `verify_apply_task` which will flush blocks after a set interval of time has elapsed as you had suggested orally. This can be controlled via the `--sync-every-n-seconds` cli arg. So now we can be certain blocks will be flushed to the db either every n seconds or every n blocks, whichever one is faster.